sdk/dotnet/Docker.cs

```diff
@@ -0,0 +1,641 @@
+    /// </summary>
+    public class CacheFrom : ResourceArgs
+    {
+        /// <summary>
+        /// An optional list of build stages to use for caching. Each build stage in this list will be 
+        /// built explicitly and pushed to the target repository.A given stage's image will be tagged as
+        /// "[stage-name]".
+        /// </summary>
+        [Input("stages")]
+        public InputList<string>? Stages { get; set; }
+    }
+
+    /// <summary>
+    /// A copy of <see cref="CacheFrom"/> but without using <see cref="Input{T}"/> in types.
+    /// </summary>
+    internal class CacheFromUnwrap
+    {
+        public ImmutableArray<string>? Stages { get; set; }
+    }
+
+    internal static class Docker
+    {
+        /// <summary>
+        /// Build and push the Dockerfile and context from a path into the requested docker 
+        /// repo.  It returns the unique target image name for the image in the docker 
+        /// repository. During preview this will build the image, and return the target image
+        /// name, without pushing.During a normal update, it will do the same, as well as tag 
+        /// and push the image.
+        /// </summary>
+        /// <param name="imageName">Docker image name.</param>
+        /// <param name="pathOrBuild">Image build parameters.</param>
+        /// <param name="repositoryUrl">Target repository URL.</param>
+        /// <param name="logResource">Associated Pulumi resource.</param>
+        /// <param name="registry">Optional image registry properties.</param>
+        /// <returns></returns>
+        public static Output<string> BuildAndPushImageAsync(string imageName, InputUnion<string, DockerBuild> pathOrBuild,
+            Input<string> repositoryUrl, Resource logResource, Input<ImageRegistry>? registry)
+        {
+            return Output.Tuple(pathOrBuild.Unwrap(), repositoryUrl, registry.Unwrap()).Apply(async v =>
+            {
+                var buildVal = v.Item1;
+                var repositoryUrlVal = v.Item2;
+                var registryVal = v.Item3;
+
+                // Give an initial message indicating what we're about to do.  That way, if anything
+                // takes a while, the user has an idea about what's going on.
+                Log.Info("Starting docker build and push...", logResource, ephemeral: true);
+
+                var result = await BuildAndPushImageWorkerAsync(
+                    imageName, buildVal, repositoryUrlVal, logResource, registryVal).ConfigureAwait(false);
+
+                // If we got here, then building/pushing didn't throw any errors.  update the status bar
+                // indicating that things worked properly.  that way, the info bar isn't stuck showing the very
+                // last thing printed by some subcommand we launched.
+                Log.Info("successfully pushed to docker", logResource, ephemeral: true);
+
+                return result;
+            });
+        }
```

Comment:
@mikhailshilkov i assume this is nearly all copy/paste.  SO i'll trust that was done pretty sanely.  Was there any particular parts you thought were relevant to get special attention for? IF so, please point it out and i'll take a closer look.

Reply:
Input unwrapping + .net specific API like process execution, but you've already checked those

Reply:
thanks.  #resolved